### DISCUSSION TITLE
EphemerisGenerator and MatricesHarvester

### DISCUSSION THREAD

**Discussion by roccafrancesco** [Hipparchus-Math/hipparchus#336] [General]
Hello everybody
i have a question on EphemerisGenerator and MatricesHarvester object;
In a case we have instantiated the two class before run numerical propagation,
what happens if we reset the numerical propagator and change the parameter drivers and run again the propagator? will the two objects be recalculated with the new propagation?

**Comment by maisonobe** [comment]
This question should rather be on the Orekit forum.
However, here the answer. Converning the `MatrixHarvester`, I think it should work properly as it does not really store states (it relies on states being passed to it at run time).  On the other hand, the ephemeris generator will gather data from both propagations, hence it will most certainly end up corrupted.
This kind of side effects is the reason why we have set up a hierarchy of `PropagatorBuilder`, they are used for example in the orbit determination classes like batch least squares. These builders allow to build a new propagator at each iteration, and to change the parameter drivers before running each propagator. The propagators are run only once.
